fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -1914,13 +1995,61 @@ protected void saveStoreHeader(@Nonnull RecordMetaDataProto.DataStoreInfo storeH
+     * is changed whenever the {@link RecordStoreState} is changed. By default, record store state information is
+     * <em>not</em> cacheable. This is because for deployments in which there are many record stores sharing the same
+     * cluster, updating the meta-data version key every time any store changes its state could lead to performance
+     * problems, so (at least for now), this is a feature the user must opt-in to on each record store.
+     *
+     * @param cacheable whether the meta-data version-stamp should be invalidated upon store state change
+     * @return a future that will complete to {@code true} if the store state's cacheability has changed
+     */
+    @Nonnull
+    public CompletableFuture<Boolean> setStateCacheabilityAsync(boolean cacheable) {
+        if (recordStoreStateRef.get() == null) {
+            return preloadRecordStoreStateAsync().thenCompose(vignore -> setStateCacheabilityAsync(cacheable));
+        }
+        if (formatVersion < CACHEABLE_STATE_FORMAT_VERSION) {
+            throw new RecordCoreException("cannot mark record store state cacheable at format version " + formatVersion);
+        }
+        if (recordStoreStateRef.get().getStoreHeader().getCacheable() == cacheable) {
+            return AsyncUtil.READY_FALSE;
+        } else {
+            return updateStoreHeaderAsync(oldHeader -> oldHeader.toBuilder()
+                    .setCacheable(cacheable)
+                    .setLastUpdateTime(System.currentTimeMillis())
+                    .build()
+            ).thenApply(ignore -> true);
+        }
+    }
+
+    /**
+     * Set whether the store state is cacheable. This operation might block if the record store state has
+     * not yet been loaded. Use {@link #setStateCacheabilityAsync(boolean)} in asychronous contexts.
+     *
+     * @param cacheable whether this store's state should be cacheable
+     * @return whether the record store state cacheability has changed
+     * @see #setStateCacheabilityAsync(boolean)
+     */
+    public boolean setStateCacheability(boolean cacheable) {
+        return context.asyncToSync(FDBStoreTimer.Waits.WAIT_SET_STATE_CACHEABILITY, setStateCacheabilityAsync(cacheable));
+    }
+
     // Actually (1) writes the index state to the database and (2) updates the cached state with the new state
     private void updateIndexState(@Nonnull String indexName, byte[] indexKey, @Nonnull IndexState indexState) {
         // This is generally called by someone who should already have a write lock, but adding them here
         // defensively shouldn't cause problems.
+        if (recordStoreStateRef.get() == null) {
+            throw new RecordCoreException("cannot update index state with a null record store state");
+        }
         beginRecordStoreStateWrite();
         try {
             context.setDirtyStoreState(true);
+            if (recordStoreStateRef.get().getStoreHeader().getCacheable()) {
```

Comment:
Is this a good place to set the metadata version stamp?

Reply:
Hm, maybe I don't understand the question, but we definitely need to update the meta-data version stamp when we update any index states as that information is cached, so if we don't, instances can keep around stale cache entries.

Reply:
Ah right. I think the fact that `getCacheable` belongs to the store header not the record store state confused me on why updating the meta-data version stamp is needed here. If it makes sense, would you want to add a parallel function to `setStateCacheability()` like `getStateCacheability()` to hide `recordStoreStateRef.get().getStoreHeader().getCacheable()`? Not a big deal anyways.